File: src/governance/temporal_replay.py
"""
Predicta Semiconductor Intelligence Platform — Phase 15 Task 1
Authoritative Temporal Replay Engine (Python)
File: src/governance/temporal_replay.py

Simulates chronological lifecycle burn-in replay across 4 discrete checkpoints:
  - Checkpoint 1 (0h): Initial screening / baseline telemetry. (No future data accessible).
  - Checkpoint 2 (24h): Early burn-in evaluation. GPR prognostic drift forecast to 168h.
  - Checkpoint 3 (96h): Intermediate verification (triggered if component routed to HOLD).
  - Checkpoint 4 (168h): Final qualification & retrospective forecast validation.

Strict Anti-Leakage & Governance Invariants:
1. Past cannot see future (strict causal temporal masking).
2. Prognostic forecast at 24h must be evaluated against ground truth only at 168h retrospective check.
3. Operating threshold (0.20) and model SHA-256 (91bb598a...) are strictly immutable.
"""

from __future__ import annotations

import copy
from typing import Any, Dict, List, Optional

from src.decision_engine.uncertainty_decision_pathway import (
    PROD_OPERATING_THRESHOLD,
    GovernedDecision,
    UncertaintyDecisionPathway,
)
from src.governance.discrimination_engine import DiscriminationEngine

CHECKPOINTS = ["0h", "24h", "96h", "168h"]


class TemporalReplayEngine:
    """Executes chronological step-by-step burn-in evaluation with temporal leakage protection."""

    def __init__(self, threshold: float = PROD_OPERATING_THRESHOLD) -> None:
        self.operating_threshold = threshold
        self.decision_pathway = UncertaintyDecisionPathway(threshold)
        self.discrimination_engine = DiscriminationEngine()

    def replay_component_lifecycle(
        self,
        full_trajectory: Dict[str, Any],
        inference_fn: Optional[Any] = None,
    ) -> Dict[str, Any]:
        """
        Replays a component's lifecycle step-by-step.

        Args:
            full_trajectory: Dictionary containing telemetry at available checkpoints
                             e.g., {'telemetry_0h': {...}, 'telemetry_24h': {...}, 'telemetry_96h': {...}, 'telemetry_168h': {...}}
            inference_fn: Optional callable (Dict[str, Any] -> Dict[str, Any]) for ML inference.
        """
        component_id = full_trajectory.get("component_id", "COMPONENT-UNKNOWN")
        lot_id = full_trajectory.get("lot_id", "LOT-UNKNOWN")
        wafer_id = full_trajectory.get("wafer_id", "WAFER-UNKNOWN")
        genealogy = full_trajectory.get("genealogy_context", {})

        history_snapshots: List[Dict[str, Any]] = []
        final_disposition = "PENDING"
        routed_to_96h_verification = False

        # --- STEP 1: Checkpoint 0h (Initial Baseline) ---
        t0_data = full_trajectory.get("telemetry_0h") or {}

        # 0h Static and baseline screening
        step_0h_eval = {
            "checkpoint": "0h",
            "evaluated_features": list(t0_data.keys()),
            "status": "PASS" if t0_data else "INSUFFICIENT_EVIDENCE",
            "decision": GovernedDecision.PASS.value if t0_data else GovernedDecision.HOLD.value,
            "message": "0h Baseline screening completed. Proceeding to 24h burn-in.",
        }
        history_snapshots.append(step_0h_eval)

        # --- STEP 2: Checkpoint 24h (Early Burn-In & Prognostics) ---
        t24_data = full_trajectory.get("telemetry_24h") or {}
        step_24h_input = {
            "checkpoint": "24h",
            "telemetry_0h": copy.deepcopy(t0_data),
            "telemetry_24h": copy.deepcopy(t24_data),
            "telemetry_96h": None,  # Strictly masked
            "telemetry_168h": None, # Strictly masked
            "component_id": component_id,
            "lot_id": lot_id,
            "wafer_id": wafer_id,
            "genealogy_context": genealogy,
        }

        # If custom inference function is provided, call it with 24h data
        inf_24 = {}
        if inference_fn is not None and callable(inference_fn) and t24_data:
            inf_input = copy.deepcopy(t24_data)
            inf_input["lot_id"] = lot_id
            inf_input["equipment_id"] = full_trajectory.get("equipment_id", "EQP-101")
            inf_24 = inference_fn(inf_input)
        else:
            # Fallback evaluation from trajectory payload
            inf_24 = full_trajectory.get("evaluation_24h") or {}

        calib_prob_24 = float(inf_24.get("probability", full_trajectory.get("calibrated_probability_24h", 0.05)))
        anomaly_ev_24 = inf_24.get("detector_evidence") or full_trajectory.get("anomaly_evidence_24h", {})
        drift_24 = inf_24.get("drift_predictions") or full_trajectory.get("drift_predictions_24h", {})
        safety_24 = inf_24.get("safety_slope") or full_trajectory.get("safety_slope_24h", {})
        discrim_24 = self.discrimination_engine.evaluate(step_24h_input)

        dec_24 = self.decision_pathway.evaluate({
            "calibrated_probability": calib_prob_24,
            "anomaly_evidence": anomaly_ev_24,
            "prognostic_evidence": drift_24,
            "safety_slope": safety_24,
            "discrimination_evidence": discrim_24,
            "ood_evidence": None,
        })

        step_24h_eval = {
            "checkpoint": "24h",
            "calibrated_probability": calib_prob_24,
            "decision": dec_24.get("decision", "MONITOR"),
            "next_action": dec_24.get("next_action"),
            "decision_factors": dec_24.get("decision_factors", []),
            "reason": dec_24.get("reason"),
            "requires_96h_verification": (dec_24.get("decision") == GovernedDecision.HOLD.value),
        }
        history_snapshots.append(step_24h_eval)

        # Check if routed to 96h verification
        if dec_24.get("decision") == GovernedDecision.HOLD.value:
            routed_to_96h_verification = True

            # --- STEP 3: Checkpoint 96h (Intermediate Verification) ---
            t96_data = full_trajectory.get("telemetry_96h") or {}
            step_96h_eval = {
                "checkpoint": "96h",
                "telemetry_available": bool(t96_data),
                "verification_status": "VERIFIED_STABLE" if t96_data.get("leakage_current", 120.0) < 180.0 else "UNSTABLE_DRIFT",
                "decision": GovernedDecision.PASS.value if t96_data.get("leakage_current", 120.0) < 180.0 else GovernedDecision.REJECT.value,
                "message": "Intermediate 96h verification evaluated following 24h HOLD routing.",
            }
            history_snapshots.append(step_96h_eval)
            if step_96h_eval["decision"] == GovernedDecision.REJECT.value:
                final_disposition = "REJECT"

        elif dec_24.get("decision") == GovernedDecision.REJECT.value:
            final_disposition = "REJECT"

        # --- STEP 4: Checkpoint 168h (Final Retrospective Qualification) ---
        t168_data = full_trajectory.get("telemetry_168h") or {}
        actual_168_failed = full_trajectory.get("actual_failed_168h", False) or (
            t168_data.get("leakage_current", 120.0) > 200.0 or t168_data.get("failed", False)
        )

        # Retrospective Prognostic Forecast Error Check
        forecast_errors = {}
        for param, drift_info in drift_24.items():
            if isinstance(drift_info, dict) and "predicted_168h" in drift_info and param in t168_data:
                pred_val = float(drift_info["predicted_168h"])
                act_val = float(t168_data[param])
                forecast_errors[param] = {
                    "predicted_168h": round(pred_val, 4),
                    "actual_168h": round(act_val, 4),
                    "absolute_error": round(abs(pred_val - act_val), 4),
                    "within_95_ci": (drift_info.get("lower_95", -1e9) <= act_val <= drift_info.get("upper_95", 1e9)),
                }

        step_168h_eval = {
            "checkpoint": "168h",
            "actual_outcome": "FAILED" if actual_168_failed else "PASSED",
            "forecast_verification": forecast_errors,
            "final_qualification_status": "FAIL" if actual_168_failed else "QUALIFIED",
        }
        history_snapshots.append(step_168h_eval)

        if final_disposition == "PENDING":
            final_disposition = "PASS" if not actual_168_failed else "REJECT"

        return {
            "component_id": component_id,
            "lot_id": lot_id,
            "wafer_id": wafer_id,
            "routed_to_96h_verification": routed_to_96h_verification,
            "final_disposition": final_disposition,
            "lifecycle_snapshots": history_snapshots,
            "temporal_leakage_audit": {
                "anti_leakage_enforced": True,
                "checkpoints_evaluated_chronologically": True,
                "future_telemetry_masked_at_step": True,
            },
        }

File: src/governance/temporal_replay.js
/**
 * Predicta Semiconductor Intelligence Platform — Phase 15 Task 1
 * Authoritative Temporal Replay Engine (JavaScript)
 * File: src/governance/temporal_replay.js
 *
 * Simulates chronological lifecycle burn-in replay across 4 discrete checkpoints:
 *   - Checkpoint 1 (0h): Initial screening / baseline telemetry. (No future data accessible).
 *   - Checkpoint 2 (24h): Early burn-in evaluation. GPR prognostic drift forecast to 168h.
 *   - Checkpoint 3 (96h): Intermediate verification (triggered if component routed to HOLD).
 *   - Checkpoint 4 (168h): Final qualification & retrospective forecast validation.
 *
 * Strict Anti-Leakage & Governance Invariants:
 * 1. Past cannot see future (strict causal temporal masking).
 * 2. Prognostic forecast at 24h must be evaluated against ground truth only at 168h retrospective check.
 * 3. Operating threshold (0.20) and model SHA-256 (91bb598a...) are strictly immutable.
 */

'use strict';

const { UncertaintyDecisionPathway, GovernedDecision, PROD_OPERATING_THRESHOLD } = require('../decision_engine/uncertainty_decision_pathway');
const { DiscriminationEngine } = require('./discrimination_engine');

const CHECKPOINTS = ['0h', '24h', '96h', '168h'];

class TemporalReplayEngine {
  constructor(threshold = PROD_OPERATING_THRESHOLD) {
    this.operatingThreshold = threshold;
    this.decisionPathway = new UncertaintyDecisionPathway(threshold);
    this.discriminationEngine = new DiscriminationEngine();
  }

  replayComponentLifecycle(fullTrajectory, inferenceFn = null) {
    const componentId = fullTrajectory.component_id || 'COMPONENT-UNKNOWN';
    const lotId = fullTrajectory.lot_id || 'LOT-UNKNOWN';
    const waferId = fullTrajectory.wafer_id || 'WAFER-UNKNOWN';
    const genealogy = fullTrajectory.genealogy_context || {};

    const historySnapshots = [];
    let finalDisposition = 'PENDING';
    let routedTo96hVerification = false;

    // --- STEP 1: Checkpoint 0h (Initial Baseline) ---
    const t0Data = fullTrajectory.telemetry_0h || {};
    const step0hEval = {
      checkpoint: '0h',
      evaluated_features: Object.keys(t0Data),
      status: Object.keys(t0Data).length > 0 ? 'PASS' : 'INSUFFICIENT_EVIDENCE',
      decision: Object.keys(t0Data).length > 0 ? GovernedDecision.PASS : GovernedDecision.HOLD,
      message: '0h Baseline screening completed. Proceeding to 24h burn-in.'
    };
    historySnapshots.push(step0hEval);

    // --- STEP 2: Checkpoint 24h (Early Burn-In & Prognostics) ---
    const t24Data = fullTrajectory.telemetry_24h || {};
    const step24hInput = {
      checkpoint: '24h',
      telemetry_0h: JSON.parse(JSON.stringify(t0Data)),
      telemetry_24h: JSON.parse(JSON.stringify(t24Data)),
      telemetry_96h: null,  // Strictly masked
      telemetry_168h: null, // Strictly masked
      component_id: componentId,
      lot_id: lotId,
      wafer_id: waferId,
      genealogy_context: genealogy
    };

    let inf24 = {};
    if (inferenceFn && typeof inferenceFn === 'function' && Object.keys(t24Data).length > 0) {
      const infInput = JSON.parse(JSON.stringify(t24Data));
      infInput.lot_id = lotId;
      infInput.equipment_id = fullTrajectory.equipment_id || 'EQP-101';
      inf24 = inferenceFn(infInput);
    } else {
      inf24 = fullTrajectory.evaluation_24h || {};
    }

    const calibProb24 = Number(inf24.probability !== undefined ? inf24.probability : (fullTrajectory.calibrated_probability_24h !== undefined ? fullTrajectory.calibrated_probability_24h : 0.05));
    const anomalyEv24 = inf24.detector_evidence || fullTrajectory.anomaly_evidence_24h || {};
    const drift24 = inf24.drift_predictions || fullTrajectory.drift_predictions_24h || {};
    const safety24 = inf24.safety_slope || fullTrajectory.safety_slope_24h || {};
    const discrim24 = this.discriminationEngine.evaluate(step24hInput);

    const dec24 = this.decisionPathway.evaluate({
      calibrated_probability: calibProb24,
      anomaly_evidence: anomalyEv24,
      prognostic_evidence: drift24,
      safety_slope: safety24,
      discrimination_evidence: discrim24,
      ood_evidence: null
    });

    const step24hEval = {
      checkpoint: '24h',
      calibrated_probability: calibProb24,
      decision: dec24.decision || 'MONITOR',
      next_action: dec24.next_action,
      decision_factors: dec24.decision_factors || [],
      reason: dec24.reason,
      requires_96h_verification: (dec24.decision === GovernedDecision.HOLD)
    };
    historySnapshots.push(step24hEval);

    // Check if routed to 96h verification
    if (dec24.decision === GovernedDecision.HOLD) {
      routedTo96hVerification = true;

      // --- STEP 3: Checkpoint 96h (Intermediate Verification) ---
      const t96Data = fullTrajectory.telemetry_96h || {};
      const leakageVal = t96Data.leakage_current !== undefined ? Number(t96Data.leakage_current) : 120.0;
      const step96hEval = {
        checkpoint: '96h',
        telemetry_available: Object.keys(t96Data).length > 0,
        verification_status: leakageVal < 180.0 ? 'VERIFIED_STABLE' : 'UNSTABLE_DRIFT',
        decision: leakageVal < 180.0 ? GovernedDecision.PASS : GovernedDecision.REJECT,
        message: 'Intermediate 96h verification evaluated following 24h HOLD routing.'
      };
      historySnapshots.push(step96hEval);
      if (step96hEval.decision === GovernedDecision.REJECT) {
        finalDisposition = 'REJECT';
      }
    } else if (dec24.decision === GovernedDecision.REJECT) {
      finalDisposition = 'REJECT';
    }

    // --- STEP 4: Checkpoint 168h (Final Retrospective Qualification) ---
    const t168Data = fullTrajectory.telemetry_168h || {};
    const actual168Failed = Boolean(
      fullTrajectory.actual_failed_168h ||
      (t168Data.leakage_current && Number(t168Data.leakage_current) > 200.0) ||
      t168Data.failed
    );

    const forecastErrors = {};
    for (const [param, driftInfo] of Object.entries(drift24)) {
      if (driftInfo && driftInfo.predicted_168h !== undefined && t168Data[param] !== undefined) {
        const predVal = Number(driftInfo.predicted_168h);
        const actVal = Number(t168Data[param]);
        const lower = driftInfo.lower_95 !== undefined ? Number(driftInfo.lower_95) : -1e9;
        const upper = driftInfo.upper_95 !== undefined ? Number(driftInfo.upper_95) : 1e9;
        forecastErrors[param] = {
          predicted_168h: Math.round(predVal * 10000) / 10000,
          actual_168h: Math.round(actVal * 10000) / 10000,
          absolute_error: Math.round(Math.abs(predVal - actVal) * 10000) / 10000,
          within_95_ci: (lower <= actVal && actVal <= upper)
        };
      }
    }

    const step168hEval = {
      checkpoint: '168h',
      actual_outcome: actual168Failed ? 'FAILED' : 'PASSED',
      forecast_verification: forecastErrors,
      final_qualification_status: actual168Failed ? 'FAIL' : 'QUALIFIED'
    };
    historySnapshots.push(step168hEval);

    if (finalDisposition === 'PENDING') {
      finalDisposition = actual168Failed ? 'REJECT' : 'PASS';
    }

    return {
      component_id: componentId,
      lot_id: lotId,
      wafer_id: waferId,
      routed_to_96h_verification: routedTo96hVerification,
      final_disposition: finalDisposition,
      lifecycle_snapshots: historySnapshots,
      temporal_leakage_audit: {
        anti_leakage_enforced: true,
        checkpoints_evaluated_chronologically: true,
        future_telemetry_masked_at_step: true
      }
    };
  }
}

module.exports = {
  TemporalReplayEngine,
  CHECKPOINTS
};
